aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -8,12 +8,84 @@
 #include <ATen/SparseCsrTensorImpl.h>
 #include <ATen/SparseCsrTensorUtils.h>
 #include <ATen/SparseTensorImpl.h>
+#include <ATen/InitialTensorOptions.h>
 
 namespace at {
 namespace native {
 
 using namespace at::sparse_csr;
 
+namespace {
+
+
+} // end anonymous namespace
+
+void _validate_sparse_csr_tensor_args(const Tensor& crow_indices, const Tensor& col_indices, const Tensor& values, IntArrayRef size) {
```

Comment:
So excellent to have the invariants collected together into one user-callable function - in the spirit of which, wouldn't we also want to check the type and device invariants (currently inlined into `set_member_tensors`) here? 

It's true that we'd want to double-checking these when a call to `_validate_sparse_csr_tensor_args()` is followed by a call to `_sparse_csr_tensor_unsafe()` (which calls `set_member_tensors()` currently), but that seems doable by factoring. 

(Aside: chunking up these invariant checks into helpers that followed the top-level structure of https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md - check_type_invariants, etc. - might make this kind of factoring more convenient, plus it'd be pretty! but maybe overkill. just a thought 😁 )

Reply:
Yeap I think the same, It'd be little overkill   :grin:

Reply:
ahaha ok but so just to be clear, it was the factoring into little helpers that seemed to me like overkill - simply adding the device and type checks to `_validate_sparse_csr_tensor_args()` (and avoiding double checking on construction) doesn't seem gratuitous, it just cleans up the API by making the user-callable invariant checker completely self-contained.

Totally understand your time constraints though - this could certainly be done in a followup by someone else. But I want to understand if there's a *design* reason not to do it that I'm missing.

Re accepting, given that I'm relatively new to the PR and its prior discussion I'll leave final signoff to @ezyang and/or @pearu, but this LGTM - also wanted to note how much @pearu's enumeration of invariants facilitates reasoning about this logic, thanks @pearu 😁

Reply:
Thanks, @bhosmer, now I understand better what u mean. I agree with the idea to have a user-callable invariant checker completely self-contained. Let me update this part.

Reply:
Thanks, @bhosmer, I really appreciate your feedback, and now I understand better what u mean. I agree with the idea to have a user-callable invariant checker completely self-contained. 

So based on this, I can move most of the checks,  except:
 
- `1.4 values.dtype == dtype`
https://github.com/pytorch/pytorch/blob/33c11299f94e5114ec18b0f1809c8c9ce0e763bb/aten/src/ATen/SparseCsrTensorImpl.cpp#L93

-  `4.4 values.device == device`
https://github.com/pytorch/pytorch/blob/33c11299f94e5114ec18b0f1809c8c9ce0e763bb/aten/src/ATen/SparseCsrTensorImpl.cpp#L116

This is because `dtype` and `device` are member variables of class `SparseCsrTensorImpl` and both are not part of the input parameters for `_validate_sparse_csr_tensor_args`.  

IMO having these checks separated in two functions `set_member_tensors` (condition 1.4 and 4.4)  `_validate_sparse_csr_tensor_args` (the rest of 1x and 4x) is not a good idea. 

In the other hand, these type invariants and device invariants checks only make sense when you are creating a new CSR sparse tensor because you are providing the preferred `dtype` and `device` for the tensor construction.   

So what do you think? Looking forward to your comments. cc @pearu

Reply:
@aocsa ah good point that the invariant situation is slightly more subtle - starting from your analysis, I think there are 
- the self-contained arg invariants, which can be checked in `_validate_sparse_csr_tensor_args_`, and 
- the invariants defined _with respect to the containing tensor_, which can be checked in `set_member_tensors`.

Looking over [the list](https://github.com/pearu/pearu.github.io/blob/main/csr_tensor_invariants.md), I think the latter includes (as you say) 1.4 - values.dtype needs to match the tensor's dtype - but also all of 4 (everybody's device needs to match the tensor's device). 

Note: I think 4.* can be usefully split into 
- a check in `_validate_sparse_csr_tensor_args_` that all subtensors have the same device, and 
- a check of just one of the subtensors against the tensor device in `set_member_tensors`.

Reply:
I agree with @bhosmer , checking the devices of all member tensors in `set_member_tensors` is suboptimal.

In fact, new_csr_tensor checks that `device` is `CPU` or `CUDA`. So, checking the devices of all member tensors can be moved to `_validate_sparse_csr_tensor_args`.

Reply:
In fact, the same pattern would apply to `dtype`: 1.1 and 1.2 can be moved to `_validate_sparse_csr_tensor_args`.

Is there any reason against moving 1.3 and 1.4 to `_validate_sparse_csr_tensor_args` as well?

Reply:
Sure, if device is restricted to CPU/CUDA  all checking the devices of all member tensors can be moved to _validate_sparse_csr_tensor_args. About dtype, I'm still thinking 1.4 should be at  `set_member_tensors`.  Unlike preferred device (2 options)  preferred dtype for the tensor construction can be  float32 (default), or float64, or int8, ..., or int64. When there is not a dtype supported, that case is caught  in the type dispatcher.